### PR TITLE
Load relationships between networks when loading migrate results to @truffle/db

### PR DESCRIPTION
**Pull request by gnidan**
This PR adds logic for the @truffle/db Project abstraction to load network genealogy info when loading migration results.

From an implementation standpoint, this defines the `generateNetworkGenealogiesLoad` generator function, which uses the `possibleAncestors` and `possibleDescendants` queries in order to identify known candidates for related networks. Given these candidates, `generateNetworkGenealogiesLoad` queries the currently-connected blockchain for historic blocks on-chain to determine if these candidates are actual matches.

To test this behavior, this PR adds use of the [fast-check](https://github.com/dubzzz/fast-check) library to do property-based testing. The testing strategy is roughly as follows:
1. Generate an arbitrary and complete model of a set of blockchains, some of which fork from each other (and some of which have completely unrelated histories)
2. Use this model to select an arbitrary list of networks to add to @truffle/db via our `generateNetworkGenealogiesLoad` function. Select these arbitrary networks in batches, where each batch represents a connection to a particular blockchain.
3. Mock a provider that interfaces with our model, to handle web3 requests inside our generator
4. Run each batch through `generateNetworksLoad`
5. Build a set of expected "latest descendant networks" (i.e., the networks that we expect @truffle/db to report as having no further descendants... the leaf nodes)
6. Query @truffle/db for all latest descendant networks and compare with what we expect.

To make all this work, this PR also adds the `ancestors` and `descendants` queries to `Network`, to return a list of ancestors/descendants for a given network. This accepts a few options for convenience, notably `onlyEarliest` for ancestors and `onlyLatest` for descendants, since sometimes we don't care about immediate relations and only want to know about the furthest relations.

Further work could be done here to test this more completely, but I propose that this further work has diminishing returns and thus I excluded it from this PR. For completeness, in addition to testing the expected "latest networks", other properties that come to mind as possible to test are:
- Test for expected earliest ancestors
- Test each network for expected ancestors and expected descendants

Of these two, the former is probably easier, but I don't think either of these will identify failure conditions that aren't identified by the existing test against expected latest descendants.

Anyway, ~300 lines of code and ~700 lines of tests later, here we go!

### CODE REVIEW

**Comment by gnidan** [comment]
Just requesting your review @fainashalts in case you wanted to take a look at my implementation here.

As discussed, this still needs tests, of course :)

**Review by fainashalts — APPROVED** [review]
LGTM! Basically all stuff we've already discussed, and the property-based testing seems particularly well-suited to this situation. Just one comment and one typo from me!

File: packages/db/src/project/migrate/networkGenealogies/index.ts [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+  // previously-known networks
+  return yield* resources.load("networkGenealogies", networkGenealogies);
+}
+
+/**
+ * Given a sparsely-populated list of artifact networks from the same
+ * blockchain, sort networks by block height and build a NetworkGenealogyInput
+ * for each pair of sequential networks.
+ */
+function collectArtifactNetworks<
```
[the remark this answers is not in the document]
Comment (gnidan) [reply packages/db/src/project/migrate/networkGenealogies/index.ts:121]: I do this on purpose :)

My reasoning is: the top of the file should have the most immediately relevant things (the exports), and then below that are internal functions.

[the remark this answers is not in the document]
Comment (gnidan) [reply packages/db/src/project/migrate/networkGenealogies/index.ts:121]: The idea being... if those functions are well-named enough to get the gist, you don't need to see those functions right away.

File: packages/db/src/project/migrate/networkGenealogies/index.ts [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+    ({
+      networks: candidates,
+      alreadyTried
+    } = yield* queryNextPossiblyRelatedNetworks(
+      relation,
+      network,
+      alreadyTried
+    ));
+
+    // check blockchain to first a matching network
```
[the remark this answers is not in the document]
Comment (gnidan) [reply packages/db/src/project/migrate/networkGenealogies/index.ts:222]: Thanks for catching this!

File: packages/db/src/project/migrate/networkGenealogies/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (gnidan) [reply packages/db/src/project/migrate/networkGenealogies/index.ts:121]: If you've ever caught me moving your code around, this is probably why :)

**Pull request merged**